build-system/server.js

```diff
@@ -79,7 +79,28 @@ app.use('/api/echo/post', function(req, res) {
   res.end(JSON.stringify(req.body, null, 2));
 });
 
+const ORIGIN_REGEX = new RegExp("http://localhost:8000|" +
+    "https?://.+\.herokuapp\.com:8000");
+const SOURCE_ORIGIN_REGEX = new RegExp("http://localhost:8000|" +
+    "https?://.+\.herokuapp\.com:8000/");
+
 app.use('/form/html/post', function(req, res) {
+  if(!ORIGIN_REGEX.test(req.headers.origin)) {
```

Comment:
My guess is that `origin` on same-origin requests can be `null`. Did you check `localhost:8000 -> localhost:8000` requests. Please also try on Chrome and Safari. Chrome may have started sending `origin` in all cases.